### DISCUSSION TITLE
Call pb_hooks event hooks from inside an event hook?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3942] [Q&A] [answered]
I'd like to do something like this:

```js
// plugin.pb.js
onAfterBootstrap(() => {
  const files = $filepath.glob([__hooks, `plugins/*`].join('/'))
  files.forEach((file) => {
    try {
      const plugin = require([file, 'index.js'].join('/'))
      plugin.init()
    } catch (e) {
      console.error(`Skipping plugin ${file}: ${e}`)
    }
  })
})
```

```js
// plugins/my-plugin/index.js
module.exports = {
   init: ()=>{
      routerAdd(`GET`, `/api/ping`, (c) => {
        return c.json(200, { message: 'Hello!' })
      })
   }
}
```

But it says ` ReferenceError: routerAdd is not defined`

**Comment by ganigeorgiev** [comment, marked as answer]
The current goja implementation works by having 2 different type of goja runtimes:
- loader (executed only once on startup to "precompile" the hooks handlers into a `goja.Program`)
- executors (executes the compiled `goja.Program` and runs every time when a hook is triggered).

Hooks can be registered only by the loader runtime (for more details see https://github.com/pocketbase/pocketbase/discussions/2825#discussioncomment-6376210).

I haven't tried it, but technically your code sample should work if you remove the `onAfterBootstrap` wrapping in the first code snippet so that the loader vm can run directly your hook registration.

**Reply by benallfree** [reply]
Oh, that worked! 💵
